### ISSUE TITLE
EntityFrameworkCore 7 SetProperty(avar,avalue) Unable to cast object of type 'System.Linq.Expressions.TypedParameterExpression' to type 'System.Linq.Expressions.LambdaExpression'.

### ISSUE THREAD

**Issue by 2366621450** [dotnet/efcore#31349]
1. Visual Studio 2022
2. .NET 7.0
3. ASP.NET Core Web API Project
4. OS Windows10
5. Microsoft.EntityFrameworkCore 7.0.9

**My code:
1.Error**
```
ADal aDal = new ADal();

Func<AModel, float> p = property => property.Balance;

aDal._databaseContext
.ModelDbSet
.Where(w => w.ID == 5)
.Take(1)
.ExecuteUpdate(e => e.SetProperty(p, 636));
```

Out error (Almost no answer found online):
引发的异常:“System.InvalidCastException”(位于 System.Private.CoreLib.dll 中)

“System.InvalidCastException”类型的未经处理的异常在 System.Private.CoreLib.dll 中发生

Unable to cast object of type 'System.Linq.Expressions.TypedParameterExpression' to type 'System.Linq.Expressions.LambdaExpression'.

**2.Normal (Do not use : Func<AModel, float> p = property => property.Balance;)**
```
ADal aDal = new ADal();

aDal._databaseContext
.ModelDbSet
.Where(w => w.ID == 5)
.Take(1)
.ExecuteUpdate(e => e.SetProperty(p => p.Balance, 636));
```

**Comment by roji** [comment]
Yeah, we should throw a more informative exception here.

**Comment by 2366621450** [comment]
I posted my question on Microsoft Q&A with detailed examples, maybe you can take a look:
https://learn.microsoft.com/en-us/answers/questions/1336727/entityframeworkcore-7-setproperty(avar-avalue)-una?comment=question

**Comment by vadimkhm** [comment]
This issue is marked as closed and solved for 8.0.0, I am using .net 8, and this issue is still present.

```
Func<User, int> p = property => property.ClientId;
DatabaseContext.Entity<User>()
               .Where(w => w.Id == 11)
               .ExecuteUpdateAsync(e => e.SetProperty(p, 30));
```

When I change it to the following it works normally:
```
DatabaseContext.Entity<User>()
               .Where(w => w.Id == 11)
               .ExecuteUpdateAsync(e => e.SetProperty(p = property => property.ClientId, 30));
```

Do you have any plans to fix this?

Thanks, Vadim

**Comment by roji** [comment]
@vadimkhm your "works normally" code is mangled, I assume you meant:

```c#
DatabaseContext.Entity<User>()
               .Where(w => w.Id == 11)
               .ExecuteUpdateAsync(e => e.SetProperty(p => p.ClientId, 30));
```

In any case, #31408 was only about generating a better exception message for most cases - this still isn't supported. Making this work would require #32018, which has already been requested for other reasons - please follow that issue.
